src/ref_count.rs

```diff
@@ -0,0 +1,370 @@
+    /// [`Ordering::Relaxed`](https://doc.rust-lang.org/std/sync/atomic/enum.Ordering.html#variant.Relaxed).
+    /// This means that reads and writes of your code may be freely reordered around this
+    /// read, there is no synchronisation with other threads.
+    ///
+    /// The value will always be at least 1. If the value is 1, this means that the interner
+    /// which produced this reference has been dropped; in this case you are still free to
+    /// use this reference in any way you like.
+    pub fn ref_count(&self) -> usize {
+        self.inner().refs.load(Relaxed)
+    }
+
+    fn inner(&self) -> &RefCounted<T> {
+        // this is safe since the existence of &self proves that the pointer is still valid
+        unsafe { self.inner.as_ref() }
+    }
+
+    pub(crate) fn from_box(value: Box<T>) -> Self {
+        Self {
+            inner: RefCounted::from_box(value),
+        }
+    }
+
+    pub(crate) fn from_sized(value: T) -> Self
+    where
+        T: Sized,
+    {
+        Self {
+            inner: RefCounted::from_sized(value),
+        }
+    }
+
+    pub(crate) fn make_hot(&mut self, map: *mut RemovePtr<T>) -> bool {
+        self.inner()
+            .remove_if_last
+            .compare_exchange(std::ptr::null_mut(), map, Relaxed, Relaxed)
+            .is_ok()
+    }
+}
+
+impl<T: ?Sized> Clone for Interned<T> {
+    fn clone(&self) -> Self {
+        const MAX_REFCOUNT: usize = usize::MAX - 1;
+        if self.inner().refs.fetch_add(1, Relaxed) == MAX_REFCOUNT {
+            // the below misspelling is deliberate
+            panic!("either you are running on an 8086 or you are leaking Interned values at a phantastic rate");
+        }
+        let ret = Self { inner: self.inner };
+        #[cfg(feature = "println")]
+        println!("clone {:p} {:p}", *self, ret);
+        ret
+    }
+}
+
+impl<T: ?Sized> Drop for Interned<T> {
+    fn drop(&mut self) {
+        // printing `self` to mark this particular execution (points to the stack)
+        // printing `*self` to mark the interned value (as printed by `clone`)
+        #[cfg(feature = "println")]
+        println!("dropping {:p} {:p}", self, *self);
+        // precondition:
+        //  - this Interned may or may not be referenced by an interner (since the interner can be dropped)
+        //  - the `self` reference guarantees that the reference count is at least one
+        //  - whatever happens, we must decrement the reference count by one
+        //  - if the only remaining reference is the interner map, we need to try to remove it
+        //    (this races against an `intern` call for the same value)
+        //
+        // IMPORTANT NOTE: each Interned starts out with two references!
+
+        // after decrementing the reference count, we must consider our memory to be freed
+        // by another thread (impossible when prior_refs == 1), so all work must happen before
+        loop {
+            if self.inner().refs.load(Relaxed) == 2 {
```

Comment:
To me it seems that it is still possible for 2 threads to drop their `Interned` at the same time, and both observing `3` here (1 from Interner + 2 live ones), leading to neither one removing the value from the Interner.

Reply:
Thanks for spotting! I found it impossible to solve this without locks, so I explicitly dropped a previously implicit constraint: what I now pushed should be correct, but it may “orphan” Interned values, making it impossible to later optimise to pointer equality — all other impls I checked have this problem, too.